### PR TITLE
A bunch of refactoring before changes in the API

### PR DESCRIPTION
**Pull request by ethouris**
As per Issue #108

### CODE REVIEW

File: srtcore/core.cpp [hunk 1]
```diff
@@ -6111,9 +6109,13 @@ void CUDT::processCtrl(CPacket& ctrlpkt)
...
-      CGuard::enterCS(m_RecvLock);
-      m_pRcvBuffer->addRcvTsbPdDriftSample(ctrlpkt.getMsgTimeStamp());
-      CGuard::leaveCS(m_RecvLock);
+      // This function will put a lock on m_RecvLock by itself, as needed.
+      // It must be done inside because this function reads the current time
+      // and if waiting for the lock has caused a delay, the time will be
+      // inaccurate. Additionally it won't lock if TSBPD mode is off, and
+      // won't update anything. Note that if you set TSBPD mode and use
+      // srt_recvfile (which doesn't make any sense), you'll have e deadlock.
+      m_pRcvBuffer->addRcvTsbPdDriftSample(ctrlpkt.getMsgTimeStamp(), m_RecvLock);
```
Comment (rndi) [inline srtcore/core.cpp]: This is a significant change/fix, especially if lock in now not taken under certain conditions. Please move to a separate pull request.

File: srtcore/buffer.cpp [hunk 2]
```diff
@@ -972,11 +1018,25 @@ bool CRcvBuffer::getRcvFirstMsg(ref_t<uint64_t> tsbpdtime, ref_t<bool> passack,
                      * Tell 1st valid packet seqno so caller can skip (drop) the missing packets.
                      */
                     skipseqno = m_pUnit[i]->m_Packet.m_iSeqNo;
+                    if ( pppkt )
+                        *pppkt = &m_pUnit[i]->m_Packet;
```
[the remark this answers is not in the document]
Comment (ethouris) [reply srtcore/buffer.cpp]: Actually this `pppkt` plays no logical role here. The use of it is only for logging. This function simply has to return some data that need to be processed further, but the packet itself is of no use to it. I've added extracting the packet just to be able to make the calling function look into its internals, in particular, extract the sequence number. Check `core.cpp` around line 3766. Previously happened that this number could be not always extracted, so the log was displaying the fallback value 0; this makes the log displaying better.

File: srtcore/buffer.cpp [hunk 3]
```diff
@@ -1389,7 +1449,7 @@ void CRcvBuffer::printDriftOffset(int tsbPdOffset, int tsbPdDriftAvg)
 }
 #endif /* SRT_DEBUG_TSBPD_DRIFT */
 
-void CRcvBuffer::addRcvTsbPdDriftSample(uint32_t timestamp)
+void CRcvBuffer::addRcvTsbPdDriftSample(uint32_t timestamp, pthread_mutex_t& mutex_to_lock)
```
[the remark this answers is not in the document]
Comment (ethouris) [reply srtcore/buffer.cpp]: Agreed, goes to a separate PR.

File: srtcore/core.cpp (hunk 1 above)
Comment (ethouris) [reply srtcore/core.cpp]: idem

File: srtcore/buffer.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (rndi) [reply srtcore/buffer.cpp]: Ok, thank you for explanation! As agreed, there will be a separate PR that will clean up declaration and usage of the packet tracing argument, changing it either to const or seq depending on the usage.

**Comment by rndi** [comment]
Thank you @ethouris !
